rocks-lib/src/upload/mod.rs

```diff
@@ -0,0 +1,177 @@
+use mlua::{Lua, Value};
+use reqwest::{
+    multipart::{Form, Part},
+    Client,
+};
+use serde::Deserialize;
+use thiserror::Error;
+
+#[derive(Deserialize, Debug)]
+pub struct VersionCheckResponse {
+    version: String,
+}
+
+#[derive(Error, Debug)]
+pub enum ToolCheckError {
+    #[error(transparent)]
+    Request(#[from] reqwest::Error),
+    #[error("`rocks` is out of date with {0}'s expected tool version! `rocks` is at version {TOOL_VERSION}, server is at {}", _1.version)]
+    ToolOutdated(String, VersionCheckResponse),
+}
+
+#[derive(Error, Debug)]
+pub enum UserCheckError {
+    #[error(transparent)]
+    Request(#[from] reqwest::Error),
+    #[error("invalid API key provided")]
+    UserNotFound,
+}
+
+#[derive(Error, Debug)]
+#[error("could not check rock status on server: {0}")]
+pub struct RockCheckError(#[from] reqwest::Error);
+
+#[derive(Error, Debug)]
+#[error(transparent)]
+pub enum UploadError {
+    Request(#[from] reqwest::Error),
+    #[error("unable to read rockspec: {0}")]
+    RockspecRead(#[from] std::io::Error),
+    Lua(#[from] mlua::Error),
+    ToolCheck(#[from] ToolCheckError),
+    UserCheck(#[from] UserCheckError),
+    RockCheck(#[from] RockCheckError),
+    #[error("rock already exists on server: {0}")]
+    RockExists(String),
+}
+
+pub async fn upload_from_project(
+    project: &Project,
+    api_key: String,
+    config: &Config,
+) -> Result<(), UploadError> {
+    let client = Client::new();
+
+    helpers::ensure_tool_version(&client, config.server()).await?;
+    helpers::ensure_user_exists(&client, &api_key, config.server()).await?;
+
+    let rockspec = project.rockspec();
+
+    if helpers::rock_exists(
+        &client,
+        &api_key,
+        &rockspec.package,
+        &rockspec.version,
+        config.server(),
+    )
+    .await?
+    {
+        return Err(UploadError::RockExists(config.server().clone()));
+    }
+
+    let rockspec_path = project.root().join("project.rockspec");
+    let rockspec_content = std::fs::read_to_string(&rockspec_path)?;
+    let lua = Lua::new();
+    lua.globals().clear()?;
+    lua.load(rockspec_content).exec()?;
+
+    let new_rockspec = {
+        let mut lines: Vec<String> = lua
+            .globals()
+            .pairs()
+            .map(|pair| {
+                let (key, value): (String, Value) = pair?;
+
+                Ok::<_, mlua::Error>(format!("{key} = {value:#?}\n"))
+            })
+            .try_collect()?;
+
+        lines.sort();
+        lines.concat()
+    };
+
+    let rockspec = Part::text(new_rockspec)
+        .file_name(format!(
+            "{}-{}.rockspec",
+            rockspec.package, rockspec.version
+        ))
+        .mime_str("application/octet-stream")?;
+    //let signature = reqwest::multipart::Part::text("something").file_name("project.rockspec.sig");
+    let multipart = Form::new().part("rockspec_file", rockspec);
+
+    client
+        .post(format!("{}{}/{}/upload", config.server(), "api/1", api_key))
```

Comment:
question(security): Can we somehow tell the client that the url contains sensitive information?

Reply:
Unsure, I'll have to look into that. That's not on us though, I don't know who thought putting the API key in the URL is a good idea :sweat_smile:

Reply:
If the client doesn't log anything or include the url in error messages, we should be fine.

> I don't know who thought putting the API key in the URL is a good idea

You raise a good point. Firewalls and proxies will likely log the URL unhashed. This is potentially a major security issue that we should raise with luarocks-site.

Reply:
Not sure if it's *major*, but it's definitely larger-than-minor. Any sort of logging immediately leaks the API key :/

Reply:
which is a risk for supply chain attacks.

Reply:
it is major omg :o

Reply:
luarocks uses https, so the path will be encrypted. But there is still a risk on the server side.